// ngx-facebook/esm5/components/fb-post/fb-post.js
import { __decorate, __extends } from "tslib";
import { Component, Input, ElementRef, Renderer2 } from '@angular/core';
import { FBMLAttribute, FBMLComponent } from '../fbml-component';
/**
 * @name Embedded Posts
 * @shortdesc Embedded post component
 * @fbdoc https://developers.facebook.com/docs/plugins/embedded-posts
 * @description
 * Embedded Posts are a simple way to put public posts - by a Page or a person on Facebook - into the content of your web site or web page.
 * Only public posts from Facebook Pages and profiles can be embedded.
 * @usage
 * ```html
 * <fb-post href="https://www.facebook.com/20531316728/posts/10154009990506729/"></fb-post>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
var FBPostComponent = /** @class */ (function (_super) {
    __extends(FBPostComponent, _super);
    function FBPostComponent(el, rnd) {
        return _super.call(this, el, rnd, 'fb-post') || this;
    }
    FBPostComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPostComponent.prototype, "href", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPostComponent.prototype, "width", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPostComponent.prototype, "showText", void 0);
FBPostComponent.ɵfac = function FBPostComponent_Factory(t) { return new (t || FBPostComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
FBPostComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FBPostComponent, selectors: [["fb-post"]], inputs: { href: "href", width: "width", showText: "showText" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function FBPostComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FBPostComponent, [{
        type: Component,
        args: [{
                selector: 'fb-post',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { href: [{
            type: Input
        }], width: [{
            type: Input
        }], showText: [{
            type: Input
        }] }); })();
    return FBPostComponent;
}(FBMLComponent));
export { FBPostComponent };
