// ngx-facebook/esm5/components/fb-page/fb-page.js
import { __decorate, __extends } from "tslib";
import { Component, Input, ElementRef, Renderer2 } from '@angular/core';
import { FBMLAttribute, FBMLComponent } from '../fbml-component';
/**
 * @name Page Plugin
 * @shortdesc Page plugin component
 * @fbdoc https://developers.facebook.com/docs/plugins/page-plugin
 * @description
 * The Page plugin lets you easily embed and promote any Facebook Page on your website. Just like on Facebook, your visitors can like and share the Page without leaving your site.
 * @usage
 * ```html
 * <fb-page href="https://facebook.com/facebook"></fb-page>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
var FBPageComponent = /** @class */ (function (_super) {
    __extends(FBPageComponent, _super);
    function FBPageComponent(el, rnd) {
        return _super.call(this, el, rnd, 'fb-page') || this;
    }
    FBPageComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "href", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "width", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "height", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "tabs", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "hideCover", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "showFacepile", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "hideCTA", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "smallHeader", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBPageComponent.prototype, "adaptContainerWidth", void 0);
FBPageComponent.ɵfac = function FBPageComponent_Factory(t) { return new (t || FBPageComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
FBPageComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FBPageComponent, selectors: [["fb-page"]], inputs: { href: "href", width: "width", height: "height", tabs: "tabs", hideCover: "hideCover", showFacepile: "showFacepile", hideCTA: "hideCTA", smallHeader: "smallHeader", adaptContainerWidth: "adaptContainerWidth" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function FBPageComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FBPageComponent, [{
        type: Component,
        args: [{
                selector: 'fb-page',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { href: [{
            type: Input
        }], width: [{
            type: Input
        }], height: [{
            type: Input
        }], tabs: [{
            type: Input
        }], hideCover: [{
            type: Input
        }], showFacepile: [{
            type: Input
        }], hideCTA: [{
            type: Input
        }], smallHeader: [{
            type: Input
        }], adaptContainerWidth: [{
            type: Input
        }] }); })();
    return FBPageComponent;
}(FBMLComponent));
export { FBPageComponent };
