// ngx-facebook/esm5/components/fb-comments/fb-comments.js
import { __decorate, __extends } from "tslib";
import { Component, Input, ElementRef, Renderer2 } from '@angular/core';
import { FBMLAttribute, FBMLComponent } from '../fbml-component';
/**
 * @name Comments
 * @shortdesc Comments component
 * @fbdoc https://developers.facebook.com/docs/plugins/comments
 * @description
 * The comments plugin lets people comment on content on your site using their Facebook account.
 * People can choose to share their comment activity with their friends (and friends of their friends) on Facebook as well.
 * The comments plugin also includes built-in moderation tools and social relevance ranking.
 *
 * @usage
 * ```html
 * <fb-comments></fb-comments>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
var FBCommentsComponent = /** @class */ (function (_super) {
    __extends(FBCommentsComponent, _super);
    function FBCommentsComponent(el, rnd) {
        var _this = _super.call(this, el, rnd, 'fb-comments') || this;
        /**
         * The absolute URL that comments posted in the plugin will be permanently associated with.
         * All stories shared on Facebook about comments posted using the comments plugin will link to this URL.
         * Defaults to current URL.
         */
        _this.href = window.location.href;
        return _this;
    }
    FBCommentsComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        FBMLAttribute
    ], FBCommentsComponent.prototype, "colorscheme", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBCommentsComponent.prototype, "href", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBCommentsComponent.prototype, "mobile", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBCommentsComponent.prototype, "numposts", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBCommentsComponent.prototype, "orderBy", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBCommentsComponent.prototype, "width", void 0);
FBCommentsComponent.ɵfac = function FBCommentsComponent_Factory(t) { return new (t || FBCommentsComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
FBCommentsComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FBCommentsComponent, selectors: [["fb-comments"]], inputs: { colorscheme: "colorscheme", href: "href", mobile: "mobile", numposts: "numposts", orderBy: "orderBy", width: "width" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function FBCommentsComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FBCommentsComponent, [{
        type: Component,
        args: [{
                selector: 'fb-comments',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { colorscheme: [{
            type: Input
        }], href: [{
            type: Input
        }], mobile: [{
            type: Input
        }], numposts: [{
            type: Input
        }], orderBy: [{
            type: Input
        }], width: [{
            type: Input
        }] }); })();
    return FBCommentsComponent;
}(FBMLComponent));
export { FBCommentsComponent };
