// ngx-facebook/esm5/components/fb-share/fb-share.js
import { __decorate, __extends } from "tslib";
import { Component, Input, ElementRef, Renderer2 } from '@angular/core';
import { FBMLAttribute, FBMLComponent } from '../fbml-component';
/**
 * @name Share Button
 * @shortdesc Share button component
 * @fbdoc https://developers.facebook.com/docs/plugins/share-button
 * @description
 * The Share button lets people add a personalized message to links before sharing on their timeline, in groups, or to their friends via a Facebook Message.
 * @usage
 * ```html
 * <fb-share href="https://github.com/zyra/ngx-facebook-sdk/"></fb-share>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
var FBShareComponent = /** @class */ (function (_super) {
    __extends(FBShareComponent, _super);
    function FBShareComponent(el, rnd) {
        return _super.call(this, el, rnd, 'fb-share-button') || this;
    }
    FBShareComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        Input(),
        FBMLAttribute
    ], FBShareComponent.prototype, "href", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBShareComponent.prototype, "layout", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBShareComponent.prototype, "mobileIframe", void 0);
    __decorate([
        Input(),
        FBMLAttribute
    ], FBShareComponent.prototype, "size", void 0);
FBShareComponent.ɵfac = function FBShareComponent_Factory(t) { return new (t || FBShareComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
FBShareComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FBShareComponent, selectors: [["fb-share"]], inputs: { href: "href", layout: "layout", mobileIframe: "mobileIframe", size: "size" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function FBShareComponent_Template(rf, ctx) { }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FBShareComponent, [{
        type: Component,
        args: [{
                selector: 'fb-share',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { href: [{
            type: Input
        }], layout: [{
            type: Input
        }], mobileIframe: [{
            type: Input
        }], size: [{
            type: Input
        }] }); })();
    return FBShareComponent;
}(FBMLComponent));
export { FBShareComponent };
