// ngx-facebook/esm2015/components/fb-send/fb-send.js
import { __decorate } from "tslib";
import { Component, Input, ElementRef, Renderer2 } from '@angular/core';
import { FBMLAttribute, FBMLComponent } from '../fbml-component';
/**
 * @name Send Button
 * @shortdesc Send button component
 * @fbdoc https://developers.facebook.com/docs/plugins/send-button
 * @description
 * The Send button lets people privately send content on your site to one or more friends in a Facebook message.
 * @usage
 * ```html
 * <fb-send href="https://github.com/zyra/ngx-facebook-sdk/"></fb-send>
 * ```
 */
import * as ɵngcc0 from '@angular/core';
let FBSendComponent = class FBSendComponent extends FBMLComponent {
    constructor(el, rnd) {
        super(el, rnd, 'fb-send');
        /**
         * The absolute URL of the page that will be sent. Defaults to the current URL.
         */
        this.href = window.location.href;
    }
};
FBSendComponent.ɵfac = function FBSendComponent_Factory(t) { return new (t || FBSendComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
FBSendComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: FBSendComponent, selectors: [["fb-send"]], inputs: { href: "href", colorScheme: "colorScheme", kidDirectedSite: "kidDirectedSite", ref: "ref", size: "size" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 0, vars: 0, template: function FBSendComponent_Template(rf, ctx) { }, encapsulation: 2 });
FBSendComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    Input(),
    FBMLAttribute
], FBSendComponent.prototype, "colorScheme", void 0);
__decorate([
    Input(),
    FBMLAttribute
], FBSendComponent.prototype, "href", void 0);
__decorate([
    Input(),
    FBMLAttribute
], FBSendComponent.prototype, "kidDirectedSite", void 0);
__decorate([
    Input(),
    FBMLAttribute
], FBSendComponent.prototype, "ref", void 0);
__decorate([
    Input(),
    FBMLAttribute
], FBSendComponent.prototype, "size", void 0);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(FBSendComponent, [{
        type: Component,
        args: [{
                selector: 'fb-send',
                template: ''
            }]
    }], function () { return [{ type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, { href: [{
            type: Input
        }], colorScheme: [{
            type: Input
        }], kidDirectedSite: [{
            type: Input
        }], ref: [{
            type: Input
        }], size: [{
            type: Input
        }] }); })();
export { FBSendComponent };
